Support for clamping replication factor

michael-redpanda opened the pull request:
<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

This PR adds a new cluster configuration parameter `minimum_topic_replications` which permits a user to
force all newly created topics to have a replication factor of at least that value.  By default this cluster config
is set to 1.

Any existing topics that have RFs less than the specified value will continue to function as normal however a WARN level
log from the cluster logging facility will be generated indicating that the specified topic has an RF less than the specified
config.  This message will be generated at start up and whenever the cluster config option is changed.

The new configuration option _does not_ have any effect on internal topics (topics outside of the `kafka` namespace or on `__audit_log`, `__consumer_offsets`, or `_schemas`).

Fixes: https://github.com/redpanda-data/core-internal/issues/947

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [X] v23.3.x
- [ ] v23.2.x
- [ ] v23.1.x

## Release Notes

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

### Improvements

* Adds new cluster configuration parameter `minimum_topic_replications` that permits Redpanda administrators to set a minimum permitted replication factor for newly created topics.

rockwotj commented on src/v/config/configuration.cc:
```diff
@@ -752,7 +754,43 @@ configuration::configuration()
+      "Minimum value permitted for a topic's replication factor",
+      {.needs_restart = needs_restart::no, .visibility = visibility::user},
+      std::nullopt,
+      {.min = 3,
```
This means that single node setups (like our docker container via `rpk container start`) will not be able to create topics?

michael-redpanda replied:
By default this config is disabled so no, it won't prevent normal operations in a single-node cluster.

graphcareful replied:
From reading the cover letter seems like by default this is not set so it has no effect, and even if it was it would only generate a WARN, is this correct?

michael-redpanda replied:
> They both need to default to 1, we'll just have to override this in cloud

I disagree here.  The purpose of this config is to prevent users of Redpanda from creating topics with an RF of 1.  This would be used in multi-node clusters to enforce the minimum to protect users data.

The default value comes in when RF is not supplied when a topic is created.

> From reading the cover letter seems like by default this is not set so it has no effect, and even if it was it would only generate a WARN, is this correct?

The WARN message only occurs for topics created _before_ the minimum value is specified.

rockwotj replied:
Ah I missed that it's a `std::optional`. LGTM then

rockwotj commented on src/v/cluster/topics_frontend.h:
```diff
@@ -189,6 +190,10 @@ class topics_frontend {
       get_partition_state(model::ntp);
 
+    /// Prints a warning log for every topic that has an RF less than the
+    /// minimum specified in minimum_topic_replication_factor
```
nit: spell out what `RF` means for newcomers, specify that minimum_topic_replication_factor is a cluster config

rockwotj commented on src/v/cluster/topics_frontend.cc:
```diff
@@ -1893,4 +1899,31 @@ topics_frontend::get_partition_state(model::ntp ntp) {
 
+void topics_frontend::print_rf_warning_message() {
+    // reduce log clutter and only run warning on shard 0
+    if (ss::this_shard_id() != 0) {
+        return;
+    }
```
This should either:
1) Go in the docs for this method
2) Be lifted to the callers.

graphcareful commented:
Question if this is the motivation:
> The lack of being able to set a minimum RF has caused issues with customers losing data because their topics are not replicated across the cluster.

and by default the new feature is disabled, how does this help? Is it mainly for cloud customers where this will be set in the  tier configs?

rockwotj reviewed (commented):
Looks pretty good, mostly just a couple of nits, after we change the min of the min to 1

graphcareful commented:
Question: Why didn't we consider just bumping `default_topic_replication` to 3? It solves the driving issue behind the change

michael-redpanda commented:
> Question: Why didn't we consider just bumping `default_topic_replication` to 3? It solves the driving issue behind the change

Wouldn't this cause issues on single-node clusters?  If you don't provide an RF when the topic is created it will use the default.  So in a single-node cluster that would attempt to create a topic with an RF of 3 (if we just simply change the default to 3)

This also doesn't prevent someone from actually setting RF=1 on new topics.

rockwotj commented on src/v/cluster/topics_frontend.cc:
```diff
@@ -1893,4 +1899,31 @@ topics_frontend::get_partition_state(model::ntp ntp) {
+        return;
+    }
+    const auto& val = _minimum_topic_replication_factor();
+    if (val.has_value()) {
+        auto min_rf = *val;
+        const auto& topics = _topics.local().topics_map();
+        for (const auto& t : topics) {
+            // Don't print warning on non-user topics
+            if (model::is_user_topic(t.first)) {
+                auto rf = t.second.get_replication_factor();
+                if (rf() < min_rf) {
+                    vlog(
+                      clusterlog.warn,
+                      "Topic {} has a replication factor less than specified "
+                      "minimum: {} < {}",
+                      t.first,
+                      rf,
+                      min_rf);
+                }
+            }
+        }
+    }
+}
```
mega nit: I find logic easier to read if it's less nested. For example I think there is less context I need to keep if this is written as follows:

```
if (!val.has_value()) {
  return;
}
auto min_rf = *val;
const auto& topics = _topics.local().topics_map();
for (const auto& [topic, config] : topics) {
  if (!model::is_user_topic(topic)) {
    continue;
  }
  if (config.get_replication_factor() >= min_rf) {
    continue;
  }
  vlog(...);
}
```

michael-redpanda replied:
Thanks.  What you're seeing is the 'evolution' of the conditional as I found more things to include/exclude.  Will update.

michael-redpanda commented:
> Looks pretty good, mostly just a couple of nits, after we change the min of the min to 1

I believe it makes sense for the min for this option to be three.  Having it be a minimum of 1 seems redundant, as the minimum RF for any topic is 1.

graphcareful commented:
> > Question: Why didn't we consider just bumping `default_topic_replication` to 3? It solves the driving issue behind the change
> 
> Wouldn't this cause issues on single-node clusters? If you don't provide an RF when the topic is created it will use the default. So in a single-node cluster that would attempt to create a topic with an RF of 3 (if we just simply change the default to 3)
> 
> This also doesn't prevent someone from actually setting RF=1 on new topics.

Right, so why can't we add some logic to just make that 1 in the case the cluster is a size of one?

rockwotj commented:
> Right, so why can't we add some logic to just make that 1 in the case the cluster is a size of one?

Because the number of nodes in a cluster is dynamic, are you suggesting a dynamic default? That seems complex and sneaky.

michael-redpanda commented:
> > > Question: Why didn't we consider just bumping `default_topic_replication` to 3? It solves the driving issue behind the change
> > 
> > 
> > Wouldn't this cause issues on single-node clusters? If you don't provide an RF when the topic is created it will use the default. So in a single-node cluster that would attempt to create a topic with an RF of 3 (if we just simply change the default to 3)
> > This also doesn't prevent someone from actually setting RF=1 on new topics.
> 
> Right, so why can't we add some logic to just make that 1 in the case the cluster is a size of one?

Even if we did, that doesn't prevent someone from creating a topic with an explicit RF of 1.

graphcareful commented:
> > > > Question: Why didn't we consider just bumping `default_topic_replication` to 3? It solves the driving issue behind the change
> > > 
> > > 
> > > Wouldn't this cause issues on single-node clusters? If you don't provide an RF when the topic is created it will use the default. So in a single-node cluster that would attempt to create a topic with an RF of 3 (if we just simply change the default to 3)
> > > This also doesn't prevent someone from actually setting RF=1 on new topics.
> > 
> > 
> > Right, so why can't we add some logic to just make that 1 in the case the cluster is a size of one?
> 
> Even if we did, that doesn't prevent someone from creating a topic with an explicit RF of 1.

If someone explicitly does that it should be allowed, after all they can also explicitly change these config vars

graphcareful commented:
> > Right, so why can't we add some logic to just make that 1 in the case the cluster is a size of one?
> 
> Because the number of nodes in a cluster is dynamic, are you suggesting a dynamic default? That seems complex and sneaky.

Agreed however we already do that with upreplication of internal topics

michael-redpanda commented:
> If someone explicitly does that it should be allowed, after all they can also explicitly change these config vars

The person creating the topic vs the person who may have admin access could be different people with different roles.  This protects the user from making a mistake and not realizing it until they have lost data.

> Agreed however we already do that with upreplication of internal topics

Correct but that's outside of the configuration system.  The configuration value for internal topic replication factor does not change based on the number of nodes in the cluster.

michael-redpanda commented on src/v/config/configuration.cc:
```diff
@@ -752,14 +754,43 @@ configuration::configuration()
+         -> std::optional<ss::sstring> {
+           if (
+             min_val.has_value()
+             && *min_val > default_topic_replication.value()) {
```
Are you referring to a situation when someone is attempting to update both configs at the same time?  Then no I don't believe so.

BenPope replied:
I am.

michael-redpanda replied:
Hmmm... thanks for bringing this up... may need to remove this as I can see this being an unintended foot gun here

graphcareful commented:
> > If someone explicitly does that it should be allowed, after all they can also explicitly change these config vars
> 
> The person creating the topic vs the person who may have admin access could be different people with different roles. This protects the user from making a mistake and not realizing it until they have lost data.
> 
I see that but there's no way to know if setting RF1 was explicit or not. Maybe the admin intends this to never happen but the user intended to do that. 

> > Agreed however we already do that with upreplication of internal topics
> 
> Correct but that's outside of the configuration system. The configuration value for internal topic replication factor does not change based on the number of nodes in the cluster.

The configuration value doesn't change but the value selected does. 1 is chosen if there's not enough nodes to support the default. We could do the same for non internal topics, albeit it does seem a bit sneaky, i agree with that point, presenting a counter argument here, these options don't seem so intuitive at first look.

michael-redpanda commented:
> The configuration value doesn't change but the value selected does. 1 is chosen if there's not enough nodes to support the default. We could do the same for non internal topics, albeit it does seem a bit sneaky, i agree with that point, presenting a counter argument here, these options don't seem so intuitive at first look.

There was a discussion between myself and @mattschumpert about whether to "silently" clamp but decided against it due to it being 'sneaky'.

graphcareful commented:
> > The configuration value doesn't change but the value selected does. 1 is chosen if there's not enough nodes to support the default. We could do the same for non internal topics, albeit it does seem a bit sneaky, i agree with that point, presenting a counter argument here, these options don't seem so intuitive at first look.
> 
> There was a discussion between myself and @mattschumpert about whether to "silently" clamp but decided against it due to it being 'sneaky'.

Agree, what about the fact that this is disabled by default? Unless people get in front of this ahead of time, know about this config option and set it, it will still occur right? Maybe it should be enabled?

BenPope commented:
> > The configuration value doesn't change but the value selected does. 1 is chosen if there's not enough nodes to support the default. We could do the same for non internal topics, albeit it does seem a bit sneaky, i agree with that point, presenting a counter argument here, these options don't seem so intuitive at first look.
> 
> There was a discussion between myself and @mattschumpert about whether to "silently" clamp but decided against it due to it being 'sneaky'.

And specifically, should you return the requested value or the actual value?
1. If you return the requested value, how do you get the actual value?
2. If you return the actual value, a topic controller might find itself in a rather long loop attempting to correct the "wrong" value.

michael-redpanda commented:
> Agree, what about the fact that this is disabled by default? Unless people get in front of this ahead of time, know about this config option and set it, it will still occur right? Maybe it should be enabled?

We will document this feature and ensure our CS/SRE people are aware of this so they can recommend it to our customers, especially those that continue to create topics with an RF of 1.

I disagree about enabling it by default (mostly due to single-node clusters).  We may enable by default for our cloud offerings, but that's a discussion for @mattschumpert

michael-redpanda commented:
@graphcareful would it be helpful to have a meeting between us and Matt to discuss your concerns?  These threads are getting long and doing this asynchronously may not be the best forum for this.

graphcareful commented:
> > Agree, what about the fact that this is disabled by default? Unless people get in front of this ahead of time, know about this config option and set it, it will still occur right? Maybe it should be enabled?
> 
> We will document this feature and ensure our CS/SRE people are aware of this so they can recommend it to our customers, especially those that continue to create topics with an RF of 1.

If this is the case we can just inform them about `default_topic_replication` and that they should set it to 3

> 
> I disagree about enabling it by default (mostly due to single-node clusters). We may enable by default for our cloud offerings, but that's a discussion for @mattschumpert

vbotbuildovich commented:
<!--RETRIED FAILURES-->
ducktape was retried in https://buildkite.com/redpanda/redpanda/builds/43063#018c8325-01b1-4824-84d7-f937b031a56a
<!--4a21fa50-3bb9-4995-8d26-ba8a7f9abf1e-->

michael-redpanda commented:
Force push `3784f66`:

* Changed name of cluster config from `minimum_topic_replication_factor` to `minimum_topic_replications`
* Updates from PR comments
* Removed 'linking' of `minimum_topic_replications` and `default_topic_replications`

vbotbuildovich commented:
<!--NEW FAILURES-->
new failures in https://buildkite.com/redpanda/redpanda/builds/43075#018c8434-11c2-4d16-a121-a03a263e0a45:
```
"rptest.tests.cluster_config_test.ClusterConfigTest.test_valid_settings"
```
<!--3e12784c-da0e-421e-855e-dc7e156efaed-->

dotnwat commented on src/v/config/configuration.cc:
```diff
@@ -753,6 +753,13 @@ configuration::configuration()
+  , minimum_topic_replication(
+      *this,
+      "minimum_topic_replications",
+      "Minimum permitted value of replication factor for new topics",
+      {.needs_restart = needs_restart::no, .visibility = visibility::user},
+      std::nullopt,
```
would it make more sense to give it a default value of 1? is it non-intuitive that the default nullopt would allow 1 (i'm guessing) but you couldn't explicitly ask for 1? seems like cloud could just set it to three in their configs?

michael-redpanda replied:
My feeling is that it's kind of redundant for it to be a min of 1.  replication factor must be a positive, odd number, which implies a minimum value of 1 at all times.

rockwotj replied:
Haha I ran into this too! Not the only one :)

michael-redpanda replied:
Had a very good conversation with Noah about this and he raised an interesting point:

Are we implying that there is no minimum value for RF when this is disabled?  Obviously the minimum for RF is 1, we have a bunch of checks in our code to ensure that, but to the customer reading the documentation does it imply that there is no minimum?  Can this cause any confusion?

Trying to ensure a consistent customer experience and to not create any confusion for them, I think I'm leaning in the direction of this _not_ being an optional and having the min be 1.

@rockwotj as you have made some comments irt to this, what are your thoughts?

rockwotj replied:
+100 to that

rockwotj commented on src/v/cluster/topics_frontend.cc:
```diff
@@ -89,7 +91,14 @@ topics_frontend::topics_frontend(
+  , _hard_max_disk_usage_ratio(hard_max_disk_usage_ratio)
+  , _minimum_topic_replication(minimum_topic_replication) {
+    _minimum_topic_replication.watch([this]() {
+        if (ss::this_shard_id() == 0) {
```
nit: only register the watch on shard 0

michael-redpanda replied:
oh fair

rockwotj commented on src/v/kafka/server/handlers/configs/config_utils.h:
```diff
@@ -320,39 +326,68 @@ struct replication_factor_must_be_positive {
+      const cluster::replication_factor& value) {
+        if (model::is_user_topic(tns)) {
+            if (auto& min_val
+                = config::shard_local_cfg().minimum_topic_replication.value();
```
Did we determine this can't race? Or what was the verdict there?

michael-redpanda replied:
I believe the issue was with the cluster configurations not the configs with alter configs (or am I not remembering something correctly?)

michael-redpanda replied:
If you're referring to the cluster configuration race condition, I went ahead and removed the linking between the two configs (which reminds me I need to update the cover sheet for this PR)

rockwotj reviewed (commented):
LGTM, a couple of suggestions/questions

michael-redpanda commented:
Force push `05e968c`:

* Registered watch on shard0 in topics_frontent

rockwotj reviewed (dismissed):
LGTM

michael-redpanda commented:
/ci-repeat 5
dt-repeat=5
skip-units
skip-redpanda-build
tests/rptest/tests/topic_creation_test.py::CreateTopicsTest.test_create_with_min_rf

michael-redpanda commented:
/ci-repeat 5
dt-repeat=5
skip-units
skip-redpanda-build
tests/rptest/tests/topic_creation_test.py::CreateTopicsTest.test_min_rf_log

michael-redpanda commented:
/ci-repeat 5
dt-repeat=5
skip-units
skip-redpanda-build
tests/rptest/tests/internal_topic_protection_test.py::InternalTopicProtectionLargeClusterTest

michael-redpanda commented:
Force push `f7fafd0`:

* Made config option not optional and minimum is now 1

vbotbuildovich commented:
/backport v23.3.x
